### ISSUE TITLE
Approach for narrow<T> is not correct

### ISSUE THREAD

**Issue by dcleblanc** [isocpp/CppCoreGuidelines#1206]
In section ES.49: If you must use a cast, use a named cast, there is the following statement:

gsl::narrow // narrow<T>(x) is static_cast<T>(x) if static_cast<T>(x) == x or it throws narrowing_error

The following sample illustrates the problem - 

unsigned int x = 0xffffffff;
if(static_cast<int>(x) == x) 
  cout << "This is safe";

The problem here is that in order to do the comparison, we cast the int back to unsigned int, and because we didn't change the bit pattern, this will evaluate as true. I haven't had the time today to apply this code to the SafeInt test suite (https://github.com/dcleblanc/SafeInt), but other conversions should also fail.

In addition, this would throw a false negative in the case of a cast to bool for all but inputs of 0 or 1.

Another problem in the code is this:

size_t y = something();
unsigned int x = narrow<unsigned int>(y);

Assuming a 32-bit int and a 32-bit memory space, this check would still have a comparison, when it should optimize away. In an x64 compile, we do need the check.

Also, there seems to be a design issue here - this seems to assume that the only problem is narrowing. Signed-unsigned mismatch can also cause problems, so while converting a negative signed short to an unsigned long may not be narrowing, it can be problematic.

I'd be happy to extract the casting portion of SafeInt to fix this.

**Comment by gdr-at-ms** [comment]
Yes, the observations are correct.  We will need to revise the implementation.  What do you think a good implementation should be?

**Comment by dcleblanc** [comment]
SafeInt has a correct implementation that is 15 years old, and used across the industry. I can easily factor out just the casting portion of it, and turn it into something we can use here. Also, it would be good to address the full range of integer overflows in GSL - I'd be happy to discuss what it would take to add SafeInt into this library (renaming would be needed, as CamelCase seems not be cool these days).

I'd also suggest renaming the cast operator to something other than narrow, because there's other problems than just narrowing. Perhaps int_cast?

In terms of just the CppCoreGuidelines documentation, it would be great to have a longer discussion around integer problems, as miscalculating buffer sizes often leads to unsafe operations, sometimes even with the use of STL or span.

**Comment by gdr-at-ms** [comment]
I suspect `narrow` and `narrow_cast` are the names we would like to retain. To be fair, if we need something that requires a full fledged library, maybe we would need to refine what it is we mean by 'narrow cast' :-).  The definition of narrowing (as defined by the language spec) indeed prohibits, among other things, implicit conversion 'signed T' to 'unsigned T' and 'unsigned T' to 'signed T'.

We are looking forward to your suggestion of better implementation.

**Comment by dcleblanc** [comment]
OK, that's fine, I just hadn't thought of casting say a negative signed short into an unsigned long as narrowing. I have a small pedantic problem with the naming, which is that in some cases, a programmer may not be aware the size of a type. As an example, Office has a number of places where various ints have been renamed, so who knows what really is under a TK or an IRUL, and whether one or the other is smaller? Or as I point out, whether it narrows can be a function of the platform. But if you like that name, OK by me.

I should be able to extract the existing code, and perhaps tidy it up a little, by sometime next week. I look forward to getting it into a review.

The implementation I have in SafeInt (from above link) starts on line 1045, and there's 10 different checks with their own logic, which brings up a question - my implementation handles casting from an enum, and I think that should be a feature of this, and it also handles casting to and from floating point, which ends up bringing in some additional headers. We may want to leave that out for now. It also special cases casting to and from bool. The implementation of the core casting code ends on line 1387.

Speaking of bool, that's another error mode of the existing check - it will never work for any input other than 0 or 1, because only those round-trip.

A nice feature is that it uses template specialization to prevent excess checks - for example, if a cast is provably OK, it works out to this:

    template < typename E >
    _CONSTEXPR14 static void CastThrow( U u, T& t ) SAFEINT_CPP_THROW
    {
        t = (T)u;
    }

I also have two ways to access the code - there is of course a throwing version, but there is also a version which returns bool, and can be used to efficiently check whether a cast is valid. The throwing versions also have a replaceable policy class that can be used to determine whether to call failfast/abort, or have a C++ exception. So should the signature for this be:

const T& narrow_cast<T, E>(U u) ?

Where E could be defaulted to say std::range_error.
